src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/VirtualProject.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Security;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+
+/// <summary>
+/// This will be replaced invoke dotnet run-api command implemented in https://github.com/dotnet/sdk/pull/48749
+/// </summary>
+internal static class VirtualCSharpFileBasedProgramProject
+{
+    /// <summary>
+    /// Adjusts a path to a file-based program for use in passing the virtual project to msbuild.
+    /// (msbuild needs the path to end in .csproj to recognize as a C# project and apply all the standard props/targets to it.)
+    /// </summary>
+    internal static string GetVirtualProjectPath(string documentFilePath)
+        => Path.ChangeExtension(documentFilePath, ".csproj");
+
+    internal static (string virtualProjectXml, bool isFileBasedProgram) MakeVirtualProjectContent(string documentFilePath, SourceText text)
+    {
+        Contract.ThrowIfFalse(PathUtilities.IsAbsolute(documentFilePath));
+        // NB: this is a temporary solution for running our heuristic.
+        // When we adopt the dotnet run-api, we need to get rid of this or adjust it to be more sustainable (e.g. using the appropriate document to get a syntax tree)
+        var tree = CSharpSyntaxTree.ParseText(text, options: CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Preview), path: documentFilePath);
+        var root = tree.GetRoot();
+        var isFileBasedProgram = root.GetLeadingTrivia().Any(SyntaxKind.IgnoredDirectiveTrivia) || root.ChildNodes().Any(node => node.IsKind(SyntaxKind.GlobalStatement));
+
+        var virtualProjectXml = $"""
```

Comment:
Skipping most of this assuming it is temporary and will be coming from the SDK?

Reply:
Yes our goal is to delete this as soon as we have onboarded the API from dotnet SDK, and basically include the editor support only when SDK is able to provide us the information we are asking for.